src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -537,7 +535,16 @@ private void WriteCompletionMessage(CompletionMessage message, Utf8JsonWriter wr
             }
             else
             {
-                JsonSerializer.Serialize(writer, message.Result, message.Result.GetType(), _payloadSerializerOptions);
+                var resultType = message.Result.GetType();
+                if (resultType == typeof(RawResult))
+                {
+                    Debug.Assert(((RawResult)message.Result).RawSerializedData.IsSingleSegment);
+                    writer.WriteRawValue(((RawResult)message.Result).RawSerializedData.First.Span, skipInputValidation: true);
```

Comment:
.First.Span? Shouldn't this write the whole thing?

Reply:
That's what the assert is for, currently we're always allocating a byte[] to store the data.